include/alp/base/internalrels.hpp

```diff
@@ -0,0 +1,796 @@
+				) {
+					return *a >= *b;
+				}
+			};
+
+			/**
+			 * This class takes a generic operator implementation and exposes a more
+			 * convenient test() function based on it. This function allows arbitrary
+			 * data types being passed as parameters, and automatically handles any
+			 * casting required for the raw operator.
+			 *
+			 * @tparam REL The generic operator implementation.
+			 *
+			 */
+			template< typename REL, enum Backend implementation = config::default_backend >
+			class RelationBase {
+
+			protected:
+
+				/** The left-hand input domain. */
+				typedef typename REL::left_type D1;
+
+				/** The right-hand input domain. */
+				typedef typename REL::right_type D2;
+
+			public:
+				/** @return Whether this relation is reflexive. */
+				static constexpr bool is_reflexive() {
+					return REL::is_reflexive;
+				}
+
+				/** @return Whether this relation is irreflexive. */
+				static constexpr bool is_irreflexive() {
+					return REL::is_irreflexive;
+				}
+
+				/** @return Whether this relation is symmetric. */
+				static constexpr bool is_symmetric() {
+					return REL::is_symmetric;
+				}
+
+				/** @return Whether this relation is antisymmetric. */
+				static constexpr bool is_antisymmetric() {
+					return REL::is_antisymmetric;
+				}
+
+				/** @return Whether this relation is transitive. */
+				static constexpr bool is_transitive() {
+					return REL::is_transitive;
+				}
+
+				/** @return Whether this relation is connected. */
+				static constexpr bool is_connected() {
+					return REL::is_connected;
+				}
+
+				/** @return Whether this relation is strongly connected. */
+				static constexpr bool is_strongly_connected() {
+					return REL::is_strongly_connected;
+				}
+
+				/**
+				 * This function checks if \f$ x REL y \f$.
+				 *
+				 * @tparam InputType1 The type of the input parameter \a x.
+				 * @tparam InputType2 The type of the input parameter \a y.
+				 *
+				 * \warning If \a InputType1 does not match \a D1 \em or \a InputType2 does
+				 *          not match \a D2, then input will be cast into temporary 
+				 *          variables of the correct types.
+				 *
+				 * \note Best performance is thus only guaranteed when all domains match.
+				 *
+				 * @param[in]  x The left-hand side input.
+				 * @param[in]  y The right-hand side input.
+				 */
+				template< typename InputType1, typename InputType2 >
+				static bool check( const InputType1 & x, const InputType2 & y ) {
+					const D1 a = static_cast< D1 >( x );
+					const D2 b = static_cast< D2 >( y );
+					return REL::check( &a, &b );
+				}
+
+				/**
+				 * This is the high-performance version of check() in the sense that no
+				 * casting is required. This version will be automatically called whenever
+				 * possible.
+				 */
+				static bool check( const D1 & x, const D2 & y ) {
+					return REL::check( &x, &y );
+				}
+			};
+
+			/**
+			 * TODO: Update for Relation
```

Comment:
Use `\TODO` or `\todo` so that it ends up in the doxy gen also.

Maybe just delete the old comments?

Reply:
Updated doxygen block.